src/app/libp2p_helper/src/libp2p_helper/app.go

```diff
@@ -64,6 +62,125 @@ func (app *app) NextId() uint64 {
 	return app.counter
 }
 
+func (app *app) AddPeers(infos ...peer.AddrInfo) {
+	app.addedPeersMutex.Lock()
+	defer app.addedPeersMutex.Unlock()
+	app._addedPeers = append(app._addedPeers, infos...)
+}
+
+// GetAddedPeers returns list of peers
+//
+// Elements of returned slice should never be modified!
+func (app *app) GetAddedPeers() []peer.AddrInfo {
```

Comment:
@georgeee I think this still needs to return a copy. The caller may not mutate the returned value, but somebody else may do so while the caller is making use of it, so it is still not safe to keep that reference around after the unlock.

Reply:
I understand your concern, but as of now list of peers `_addedPeers` is being handled in such a way that elements are never modified. Slice that is written to `_addedPeers` can be replaced with another slice (like in case of appending elements to this list), but elements are never overwritten, which makes use of this function safe as long as caller respects the verbal contract of the doc comment.

Reply:
@georgeee Elements being overwritten is not the only risk here, the container itself might get corrupted.
I'm not familiar with `go` but from [this doc](https://go.dev/blog/slices-intro), this:
```
_addedPeers:              []peer.AddrInfo{},
```
is just a pointer to an array with its length and capacity, so 3 values: `[ptr; len; cap]`.

Here the function `GetAddedPeers()` returns those 3 values, but do you know if it's a reference to those 3 values, or a copy of them ?
- If it copies the 3 values, it's fine
- In case it's a reference, it will get corrupted when another thread executes one of those lines:
`app._addedPeers = append(app._addedPeers, infos...)`
`app._addedPeers = nil`
1 thread might modify `ptr`, while another thread already have read the old `len` and `cap`

It's probably safer for `GetAddedPeers` to return a copy

Reply:
These three values will be copied by value in Go. All values are passed by value in Go. There are pointer types, but in case of slice type is just these three values that you mentioned.

There is an interesting case of `append` operation, but it's also fine, you may check some [random thread](https://www.reddit.com/r/golang/comments/m5jd99/maybe_a_stupid_question_about_appending_to_slice/#:~:text=The%20resulting%20value%20of%20append,to%20the%20newly%20allocated%20array.) on Internet which I used to double-confirm my understanding of the model behind slice type.

If you have doubts, it's easy to write an app that would test this assertion, but I doubt it's necessary, from what I read in Go docs and how slices are used in Golang's std library.

Reply:
@georgeee  Ah cool, if it's a copy of those 3 values it's good